src/plan_add_hashagg.c

```diff
@@ -0,0 +1,595 @@
+} CustomEstimateForFunctionInfoHashEntry;
+
+/* Definitions of functions with a custom group estimate.
+   These will also be entered into the custom_estimate_func_hash hash table. */
+static CustomEstimateForFunctionInfo custom_estimate_func_info[] =
+{
+	{
+		.extension_function = true,
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, TIMESTAMPOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket
+	},
+	{
+		.extension_function = true,
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, TIMESTAMPTZOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket,
+	},
+	{
+		.extension_function = true,
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, DATEOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket,
+	},
+	{
+		.function_name = "date_trunc",
+		.nargs = 2,
+		.arg_types = {TEXTOID, TIMESTAMPOID},
+		.custom_group_estimate_func = custom_group_estimate_date_trunc,
+	},
+	{
+		.nargs = 2,
+		.custom_group_estimate_func = custom_group_estimate_date_trunc,
+		.arg_types = {TEXTOID, TIMESTAMPTZOID},
+		.function_name = "date_trunc",
+	},
+};
+#define _MAX_HASHAGG_FUNCTIONS (sizeof(custom_estimate_func_info)/sizeof(custom_estimate_func_info[0]))
+
+HTAB	   *custom_estimate_func_hash = NULL;
+
+static bool
+function_types_equal(Oid left[], Oid right[], int nargs)
+{
+	int			arg_index;
+
+	for (arg_index = 0; arg_index < nargs; arg_index++)
+	{
+		if (left[arg_index] != right[arg_index])
+			return false;
+	}
+	return true;
+}
+
+static void
+initialize_custom_estimate_func_info()
+{
+	int			i = 0;
+	Oid			extension_schema = extension_schema_oid();
+	char	   *extension_schema_name = get_namespace_name(extension_schema);
+
+
+	HASHCTL		hashctl = {
+		.keysize = sizeof(Oid),
+		.entrysize = sizeof(CustomEstimateForFunctionInfoHashEntry)
+	};
+
+	custom_estimate_func_hash = hash_create("custom_estimate_func_hash", _MAX_HASHAGG_FUNCTIONS, &hashctl, HASH_ELEM | HASH_BLOBS);
+
+	for (i = 0; i < _MAX_HASHAGG_FUNCTIONS; i++)
+	{
+		CustomEstimateForFunctionInfo def = custom_estimate_func_info[i];
+		CustomEstimateForFunctionInfoHashEntry *hash_entry;
+		bool		function_found = false;
+		bool		hash_found;
+		FuncCandidateList funclist;
+
+		if (def.extension_function)
+			funclist = FuncnameGetCandidates(list_make2(makeString(extension_schema_name), makeString(def.function_name)),
+											 def.nargs, NIL, false, false, false);
+		else
+			funclist = FuncnameGetCandidates(list_make1(makeString(def.function_name)),
+											 def.nargs, NIL, false, false, false);
+
+		/* check types */
+		while (!function_found && funclist != NULL)
+		{
+			if (funclist->nargs != def.nargs || !function_types_equal(funclist->args, def.arg_types, def.nargs))
+				funclist = funclist->next;
+			else
+				function_found = true;
+		}
+
+		if (!function_found)
+			elog(ERROR, "oid lookup failed for the function %s with %d args",
```

Comment:
Should `oid` be `OID` in this error message? Otherwise I think the "standard" PG error for this is: 

```
elog(ERROR, "cache lookup failed for function %u", funcoid);
```